### PR TITLE
refactor: remove "node" type dependency from client runtime

### PR DESCRIPTION
**Pull request by kt3k**
This PR removes 'node.js' type dependency from `runtime/index.d.ts`

There were 2 node.js type dependencies in client runtime.

- `sql-template-tag@4.0.0` depends on Node.js `util`, which was removed in v4.1.0
- `QueryEngineConfig["env"]` and `GetConfigOptions["env"]` use `NodeJS.ProcessEnv` type.

This PR removes the above 2 deps by upgrading `sql-template-tag` and replacing `NodeJS.ProcessEnv` with `Record<string, string | undefined>`

---
*motivations*

This is a preliminary work for Deno support. Deno doesn't have 'node' types. This makes `runtime/index.d.ts` usable as is in deno.
I believe this change is also desirable for other edge runtimes as they don't provide node.js modules by default

### CODE REVIEW

**Comment by Jolg42** [comment]
I'll let @aqrln review, note from the past
We mentioned in https://github.com/prisma/prisma/pull/13823 that we want to `We want to upgrade directly to 5.x later.`, I'm not sure why. Maybe it's fine to do the upgrade to `4.1.0` here.

**Comment by kt3k** [comment]
`sql-template-tag` dropped `object` type from `RawValue` type (see https://github.com/blakeembrey/sql-template-tag/commit/d1fb92ec5653a5680f0bf9abfbecd4518b56bef4). `RawValue` type is used as parameter of `sql` tagged template literal. That causes type errors in the usages like the below:

```ts
Prisma.sql`INSERT INTO "Entry" ("id", "binary") VALUES ('4', ${Buffer.from('hello')})`
```
We need to change this, for example, like the below to avoid type error.
```ts
Prisma.sql`INSERT INTO "Entry" ("id", "binary") VALUES ('4', ${'hello'})`
```

Is this change acceptable? (If not, I'll try `sql-template-tag@5.0.3`, which allow `unknown` type for `RawValue` type.)

**Comment by kt3k** [comment]
> Prisma.sql`INSERT INTO "Entry" ("id", "binary") VALUES ('4', ${Buffer.from('hello')})`

It looks like these Buffer values actually need to be Buffer type. I'll revert the test changes and try to upgrade `sql-template-tag` to v5.0.3.

**Comment by aqrln** [comment]
> I'll let @aqrln review, note from the past
We mentioned in https://github.com/prisma/prisma/pull/13823 that We want to upgrade directly to 5.x later., I'm not sure why. Maybe it's fine to do the upgrade to 4.1.0 here.

Sorry for not responding, I was off this Monday and Tuesday.

From the comments and commits it looks like @kt3k has already figured this out (thank you!), but just to confirm and provide additional context: `sql-template-tag@4.1.0` introduced an accidental breaking change that affected us, when we noticed that (which took quite a while because our renovate pipeline was stalled), the author immediately responded to us and released a fix in 5.0.3. However we couldn't upgrade to 5.0.3 because of https://github.com/prisma/prisma/issues/7618.

Here's the relevant PR: https://github.com/prisma/prisma/pull/11506

Adding it to `forceTranspile` like in https://github.com/prisma/prisma/pull/14589/commits/bcc492724c020257dd91ceb5175710034d95f859 might do the trick and make it work, if my understanding is correct (@millsp knows more about it).

File: packages/client/src/runtime/getPrismaClient.ts
Comment (aqrln) [inline packages/client/src/runtime/getPrismaClient.ts:57]: Maybe `import sqltag, { Sql } from 'sql-template-tag'` instead on line 28 and replace `sqlTemplateTag.Sql` with just `Sql` in the file

File: packages/client/src/utils/getTestClient.ts
Comment (aqrln) [inline packages/client/src/utils/getTestClient.ts:52]: Was the type annotation removed just because it wasn't necessary or because it was causing some problems?

Comment (kt3k) [reply packages/client/src/utils/getTestClient.ts:52]: I changed the reference to runtime module from `../runtime/` (`client/src/runtime`) to `../../runtime/` (`client/runtime`) in the above lines, and `client/runtime` doesn't export `GetPrismaClientConfig`. So I removed `GetPrismaClientConfig`. (also it doesn't look absolutely necessary)

I changed the reference to runtime like this because if we import from `client/src/runtime/` it throws require error at `sql-template-tag` because of ESM issue. The built version of runtime is in CommonJS, so it can avoid that issue. I'm not 100% sure this approach is acceptable or not. What do you think?

File: packages/client/src/runtime/getPrismaClient.ts
Comment (kt3k) [reply packages/client/src/runtime/getPrismaClient.ts]: Good idea! Updated.

File: packages/client/src/utils/getTestClient.ts
Comment (kt3k) [reply packages/client/src/utils/getTestClient.ts:52]: Ah, this change seems causing the mismatch of query engine version in CI

I now see these errors..
```
Query engine library for current platform "debian-openssl-1.1.x" could not be found.
    You incorrectly pinned it to debian-openssl-1.1.x

    This probably happens, because you built Prisma Client on a different platform.
    (Prisma Client looked in "/home/runner/work/prisma/prisma/packages/client/runtime/libquery_engine-debian-openssl-1.1.x.so.node")
```

**Comment by kt3k** [comment]
@aqrln Thank you for the feedbacks!

> However we couldn't upgrade to v5 because of https://github.com/prisma/prisma/issues/7618.

This problem seems very complex to me. I've been trying a few workarounds for this, but I still see the errors..

> Adding it to forceTranspile like in https://github.com/prisma/prisma/commit/bcc492724c020257dd91ceb5175710034d95f859 might do the trick and make it work, if my understanding is correct, but I'm not sure (@millsp knows more about it).

This seems to solve the issue for the case when sql-template-tag is directly imported in jest environment.

But there are other cases where test case invokes another node.js process (e.g. `packages/client/src/__tests__/integration/happy/signals/test.ts`), and that process imports sql-template-tag, which still causes error while importing it.

To avoid that one, I did [this change](https://github.com/prisma/prisma/pull/14589/commits/92277d445893b89a6ba3f5dc77fff33f4e5db0f3), and now I see another error about version mismatch of query engines..

**Comment by kt3k** [comment]
In what path does the query engine exist in CI?

The error message says it couldn't find it in the following paths, for example:
```
/home/ubuntu/actions-runner/_work/prisma/prisma/.prisma/client
/home/ubuntu/actions-runner/_work/prisma/prisma/packages/client/runtime
/home/ubuntu/actions-runner/_work/prisma/prisma/packages/client
/home/ubuntu/actions-runner/_work/prisma/prisma/packages/client/src/__tests__/integration/happy/insensitive-postgresql-feature-flag
/home/ubuntu/actions-runner/_work/prisma/prisma/packages/client/src/__tests__/integration/happy/insensitive-postgresql-feature-flag
/tmp/prisma-engines
/home/ubuntu/actions-runner/_work/prisma/prisma/packages/client/src/__tests__/integration/happy/insensitive-postgresql-feature-flag
```
But on my local machine, the query engine exists at `packages/client/libquery_engine-darwin.dylib.node`. So it feels strange that it couldn't find it from `packages/client`..

**Comment by millsp** [comment]
> In what path does the query engine exist in CI?

The engine is copied in two locations, directly into `@prisma/engines` on postinstall, which has also the side-effect to put a copy into the cache (if possible in eg. `/home/.cache/prisma/...`. Then there is a third location for our new `packages/client/tests/functional` (which is an effort to port tests from `__tests__/integration`), in this case the engine is also copied to the root of `packages/client`. Which tests are failing?

**Comment by kt3k** [comment]
Thanks for the explanation!

> Which tests are failing?

Some examples of failing tests are:

- `packages/client/src/__tests__/integration/happy/rejectNotFound/test.ts`
- `packages/integration-tests/src/__tests__/integration/sqlite/runtime.test.ts`
- `packages/integration-tests/src/__tests__/integration/postgresql/runtime.test.ts`
- `packages/integration-tests/src/__tests__/integration/mariadb/runtime.test.ts`

**Comment by kt3k** [comment]
Having looked at details of current test failures, the number of failures are overwhelming. I think it's not a good idea to to do workaround for `sql-template-tag@5.0.3`. Especially this commit [92277d4](https://github.com/prisma/prisma/pull/14589/commits/92277d445893b89a6ba3f5dc77fff33f4e5db0f3) looks bad

I'd rather try type casting with `sql-template-tag@4.1.0`, as the problem of it is only about typing

**Comment by aqrln** [comment]
> I'd rather try type casting with sql-template-tag@4.1.0, as the problem of it is only about typing

The problem is we're re-exporting the template tag from that library. We could work around it for ourselves with type casting, but our users' code would still be broken with that version.

**Comment by kt3k** [comment]
@millsp @aqrln I think this PR is now ready for review. PTAL

File: packages/client/src/utils/getTestClient.ts
Comment (millsp) [reply packages/client/src/utils/getTestClient.ts]: It could have been a viable option to call `generateInFolder` instead, that might have solved the issue. I feel like using the runtime as it is shipped (like you proposed) is a viable option. The tests in `__tests__` are currently being upgraded to a cleaner and more standard setup, so it is only a matter of time until we remove them. Additionally, swapping `getPrismaClient` to `generateInFolder` should be transparent.

File: packages/client/src/runtime/utils/sql-template-tag.ts
Comment (millsp) [inline packages/client/src/runtime/utils/sql-template-tag.ts:9]: This looks good although I prefer the alternative with `generateInFolder`, if you have time and if it actually works out that easily. This seems to be viable, but of course it adds that thin layer between us and the library, which I'd like to avoid if possible. If not, `__tests__` are being phased out, and we can add a TODO as a comment here that when we don't depend on `getTestClient` anymore, we can basically upgrade to v5 since we bundle to CJS.

File: packages/client/src/utils/getTestClient.ts
Comment (millsp) [reply packages/client/src/utils/getTestClient.ts]: (see https://github.com/prisma/prisma/pull/14589#discussion_r947135925)

File: packages/client/src/runtime/utils/sql-template-tag.ts
Comment (kt3k) [reply packages/client/src/runtime/utils/sql-template-tag.ts:9]: Ok, I'll experiment with `generateInFolder` for a bit.

File: packages/integration-tests/src/__tests__/__helpers__/integrationTest.ts
Comment (kt3k) [inline packages/integration-tests/src/__tests__/__helpers__/integrationTest.ts:242]: Changed this `getTestClient` call to `generateInFolder`. This has currently 2 problems:

- The test fails with OOM error (if I reduces the number of scenarios, then they pass.)
- The assertions about Decimal don't pass because the Decimal constructors don't match

File: packages/client/src/runtime/utils/sql-template-tag.ts
Comment (millsp) [reply packages/client/src/runtime/utils/sql-template-tag.ts]: If it is too tricky to get the integration-tests to pass, just leave it as is and ping me.

File: packages/integration-tests/src/__tests__/__helpers__/integrationTest.ts
Comment (millsp) [reply packages/integration-tests/src/__tests__/__helpers__/integrationTest.ts:242]: I see, let's keep your original work, I hoped this would not get too tricky. Thanks for trying!

Comment (millsp) [reply packages/integration-tests/src/__tests__/__helpers__/integrationTest.ts:242]: (these tests are also legacy tests that we will eventually get rid of)

Comment (kt3k) [reply packages/integration-tests/src/__tests__/__helpers__/integrationTest.ts:242]: Ok, I'll move this back to `sql-template-tag@4.1.0 + type modification` version

File: packages/client/src/runtime/utils/sql-template-tag.ts
Comment (millsp) [inline packages/client/src/runtime/utils/sql-template-tag.ts:5]: Awesome, thank you.

**Review by aqrln — APPROVED** [review]
Nice work, thank you!

**Comment by millsp** [comment]
Hey @kt3k, I have taken some time to upgrade to `sql-template-tag@5` via this [PR](https://github.com/prisma/prisma/pull/14885). I'm waiting for reviews on it, yours is also welcome :)

**Comment by kt3k** [comment]
@millsp Great! Thank you!

I'll update this PR soon.

**Comment by kt3k** [comment]
Updated. Now this PR only has `NodeJS.ProcessEnv | Record<string, string>` -> `Record<string, string | undefined>` update and the test of the runtime declaration file.

**Review by aqrln — APPROVED** [review]
LGTM!

**Review by millsp — APPROVED** [review]
Thank you for this PR and your efforts to get Prisma working with Deno :sauropod: :rocket:

**Pull request merged**
